code/game/objects/items/sec_lawplanner.dm

```diff
@@ -0,0 +1,367 @@
+	var/start_timer = FALSE //If true, automatically start the timer on upload
+	var/time_arrest = FALSE //If true, start counting time when the arrest is made, to subtract from the sentence.
+	var/timing = 0	//Time of arrest.
+	var/datum/data/record/upload_crimes = null //If has DNA, will look for an associated datacore file and upload crimes
+	var/list/rapsheet = list()
+	var/total_time = 0
+
+/obj/item/device/law_planner/attack_self(mob/user)
+	ui_interact(user)
+
+/obj/item/device/law_planner/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open=NANOUI_FOCUS)
+	if(user.stat && !isAdminGhost(user))
+		return
+
+	// this is the data which will be sent to the ui
+	var/data[0]
+	data["timer"] = total_time
+	data["announce"] = announce
+	data["starttimer"] = start_timer
+	data["timearrest"] = time_arrest
+	data["arresttime"] = timing
+	if(upload_crimes)
+		data["perp"] = upload_crimes.fields["name"]
+	data["crimes"] = english_list(rapsheet)
+
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, data, force_open)
+	if (!ui)
+		ui = new(user, src, ui_key, "lawplanner.tmpl", "Law Planning Frame", 520, 500)
+		ui.set_initial_data(data)
+		ui.open()
+		ui.set_auto_update(1)
+
+/obj/item/device/law_planner/Topic(href, href_list)
+	if(..(href, href_list))
+		return
+
+	if(href_list["close"])
+		if(usr.machine == src)
+			usr.unset_machine()
+		return 1
+
+	var/datum/law/L
+	if(href_list["assault"])
+		L = findlaw(LAW_ASSAULT,text2num(href_list["assault"]))
+	else if(href_list["damage"])
+		L = findlaw(LAW_DAMAGE,text2num(href_list["damage"]))
+	else if(href_list["theft"])
+		L = findlaw(LAW_THEFT,text2num(href_list["theft"]))
+	else if(href_list["contraband"])
+		L = findlaw(LAW_CONTRABAND,text2num(href_list["contraband"]))
+
+
+	switch(href_list["toggle"])
+		if("announce")
+			announce = !announce
+		if("starttimer")
+			start_timer = !start_timer
+		if("timearrest")
+			time_arrest = !time_arrest
+
+	switch(href_list["clear"])
+		if("record")
+			upload_crimes = null
+		if("rapsheet")
+			rapsheet.Cut()
+			total_time = 0
+		if("arresttime")
+			timing = 0
+
+	if(!L)
+		updateUsrDialog()
+		return
+	rapsheet += initial(L.name)
+	total_time += initial(L.penalty)
+	if(initial(L.death))
+		visible_message("[bicon(src)] <B>\The [src]</B> beeps, \"This inmate is eligible for execution.\"")
+
+	if(initial(L.demotion))
+		visible_message("[bicon(src)] <B>\The [src]</B> beeps, \"This inmate is eligible for demotion.\"")
+	updateUsrDialog()
+
+
+/obj/item/device/law_planner/proc/announce()
+	say(english_list(rapsheet))
+	say("[total_time] minutes.")
+
+/obj/item/device/law_planner/preattack(var/atom/A, var/mob/user, var/proximity_flag)
+	if(!proximity_flag)
+		return 1
+	if(!allowed(user))
+		to_chat(user, "<span class='warning'>You must wear your ID!</span>")
+		return 1
+	if(ishuman(A)&&!(A==user))
+		for(var/datum/data/record/E in data_core.security)
+			if(E.fields["name"] == A.name)
```

Comment:
This check should probably be more robust than comparing directly to the human's `name` var. For instance this will fail completely if they're wearing someone else's ID, rather than returning the record for either the ID or the person.

Reply:
In fact, the comment on `upload_crimes` seems to indicate this would be based on the DNA, not the name at all.